File: 2615-sum-of-distances/solution.py
from collections import defaultdict
from typing import List

class Solution:
    def distance(self, nums: List[int]) -> List[int]:
        n = len(nums)
        ans = [0] * n
        # Map each number to the list of indices where it appears
        groups = defaultdict(list)
        for i, val in enumerate(nums):
            groups[val].append(i)
            
        for val in groups:
            indices = groups[val]
            k = len(indices)
            if k <= 1:
                continue
            
            # total_sum is used to calculate the sum of indices to the right efficiently
            total_sum = sum(indices)
            left_sum = 0
            
            for m, idx in enumerate(indices):
                # Number of elements to the left of indices[m] is m
                # Number of elements to the right of indices[m] is (k - 1 - m)
                right_sum = total_sum - left_sum - idx
                
                # Sum of (idx - left_indices)
                left_part = m * idx - left_sum
                # Sum of (right_indices - idx)
                right_part = right_sum - (k - 1 - m) * idx
                
                ans[idx] = left_part + right_part
                
                # Update prefix sum for the next index in the group
                left_sum += idx
                
        return ans

File: 2615-sum-of-distances/solution.cpp
#include <vector>
#include <unordered_map>

using namespace std;

class Solution {
public:
    vector<long long> distance(vector<int>& nums) {
        int n = nums.size();
        vector<long long> result(n, 0);
        unordered_map<int, vector<int>> groups;

        // Group indices by their values
        for (int i = 0; i < n; ++i) {
            groups[nums[i]].push_back(i);
        }

        // Process each group of indices
        for (auto const& [val, indices] : groups) {
            int k = indices.size();
            if (k <= 1) continue;

            // Calculate total sum of indices for suffix calculations
            long long total_sum = 0;
            for (int idx : indices) {
                total_sum += idx;
            }

            long long left_sum = 0;
            for (int m = 0; m < k; ++m) {
                long long current_idx = indices[m];
                
                // Indices to the left: m elements
                // Indices to the right: (k - 1 - m) elements
                long long right_sum = total_sum - left_sum - current_idx;
                
                long long left_part = (long long)m * current_idx - left_sum;
                long long right_part = right_sum - (long long)(k - 1 - m) * current_idx;
                
                result[current_idx] = left_part + right_part;
                
                left_sum += current_idx;
            }
        }

        return result;
    }
};